### PR TITLE
Chunk worker

### PR DESCRIPTION
**Pull request by mickamcia**
Generates and removes chunks in proximity.
Visible chunk seams no longer exist.
Mining and building close to a chunk seam works properly.

Has some problems with physics engine collision mesh removal. (removal not yet implemented).
Mining and building is poorly performing.

### CODE REVIEW

File: Hyper/MarchingCubes/ChunkWorker.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                        // chunk?.DisposeCollisionSurface(_simulationManager.Simulation, _simulationManager.BufferPool);
+                    }
+                });
+                Thread.Sleep(100);
+            }
+        }
+
+        public static void UpdateNeededChunksBasedOnPosition(ChunkWorker chunkWorker, Vector3 position)
+        {
+            const int chunkCreateRadius = 50;
```
Comment (KarolDenst) [inline Hyper/MarchingCubes/ChunkWorker.cs:78]: IF i change these to 150 and 170 there is a problem. Then i cross a chunk boarder I get a lag spike for about 0.5 seconds. I think its bacaouse ChunkWorker eats all the resources. Maybe limit the number of threads or something so that the loading takes a little bit longer but we do not get lag spikes

File: Hyper/MarchingCubes/ChunkWorker.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+using Hyper.MarchingCubes;
+using Hyper.Meshes;
+using OpenTK.Mathematics;
+using Hyper.Collisions.Bepu;
+
+namespace Hyper
+{
+    internal class ChunkWorker
+    {
+        private ConcurrentDictionary<Vector3i, Guid> _neededChunkPositions;
```
Comment (KarolDenst) [inline Hyper/MarchingCubes/ChunkWorker.cs:16]: All of those could be made readonly

File: Hyper/MarchingCubes/ChunkWorker.cs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        private ConcurrentDictionary<Vector3i, Guid> _existingChunkPositions;
```
Comment (KarolDenst) [inline Hyper/MarchingCubes/ChunkWorker.cs:17]: I don't get why we use GUID when Vector3i are unique and we could just use them

File: Hyper/MarchingCubes/ChunkWorker.cs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                {
+                    if (!_neededChunkPositions.ContainsKey(kvp.Key))
+                    {
+                        _existingChunkPositions.TryRemove(kvp.Key, out _);
+                        _existingChunks.TryRemove(kvp.Value, out var chunk);
+                        // multithreaded removal might need a mutex?
+                        // chunk?.DisposeCollisionSurface(_simulationManager.Simulation, _simulationManager.BufferPool);
+                    }
+                });
+                Thread.Sleep(100);
```
Comment (KarolDenst) [inline Hyper/MarchingCubes/ChunkWorker.cs:72]: Why do we need a sleep? Maybe call this on every update frame or do something else to avoid wasting a thread

File: Hyper/Meshes/Chunk.cs [hunk 5]
```diff
@@ -25,8 +25,14 @@ internal class Chunk : Mesh
 
     private TypedIndex _shape;
 
+    private BepuPhysics.Collidables.Mesh? collisionSurface;
+    public Vector3i Center
+    {
+        get { return new Vector3i(Position.X + Size / 2, Position.Y + Size / 2, Position.Z + Size / 2); }
+    }
```
Comment (KarolDenst) [inline Hyper/Meshes/Chunk.cs:32]: add empty line

Comment (KarolDenst) [reply Hyper/Meshes/Chunk.cs:32]: public Vector3i Center => new(Position.X + Size / 2, Position.Y + Size / 2, Position.Z + Size / 2);

File: Hyper/Meshes/Chunk.cs [hunk 6]
```diff
@@ -58,17 +68,19 @@ public bool Mine(Vector3 location, float deltaTime, float brushWeight, int radiu
                     if (DistanceSquared(x, y, z, xi, yi, zi) <= radius * radius)
                     {
                         _voxels[xi, yi, zi].Value += deltaTime * brushWeight * Gaussian(xi, yi, zi, x, y, z, 0.1f);
+                        flag = true;
```
Comment (KarolDenst) [inline Hyper/Meshes/Chunk.cs:71]: flag is a bad name

File: Hyper/Meshes/Chunk.cs [hunk 7]
```diff
@@ -98,36 +109,49 @@ public bool Build(Vector3 location, float deltaTime, float brushWeight, int radi
...
+        {
+            CreateCollisionSurface(simulation, bufferPool);
+            return;
+        }
         simulation.Shapes.RemoveAndDispose(_shape, bufferPool);
         var mesh = MeshHelper.CreateMeshFromChunk(this, bufferPool);
         _shape = simulation.Shapes.Add(mesh);
         simulation.Statics[_handle].SetShape(_shape);
     }
+    public void DisposeCollisionSurface(Simulation simulation, BufferPool bufferPool)
```
Comment (KarolDenst) [inline Hyper/Meshes/Chunk.cs:138]: this is never used

File: Hyper/Meshes/Mesh.cs [hunk 8]
```diff
@@ -10,6 +10,9 @@ internal class Mesh : IDisposable
 {
     public Vector3 Position { get; set; }
 
+    public Vertex[] vertices;
```
Comment (KarolDenst) [inline Hyper/Meshes/Mesh.cs:13]: Vertices or private _vertices

File: Hyper/Meshes/Mesh.cs [hunk 9]
```diff
@@ -20,18 +23,27 @@ internal class Mesh : IDisposable
 
     protected int NumberOfVertices;
 
+    private bool createdVertexArrayObject;
```
Comment (KarolDenst) [inline Hyper/Meshes/Mesh.cs:26]: _

Comment (KarolDenst) [reply Hyper/Meshes/Mesh.cs:26]: also maybe change int vaoId -> int? vaoId and just check if its null instead of this

File: Hyper/Meshes/Mesh.cs [hunk 10]
```diff
@@ -20,18 +23,27 @@ internal class Mesh : IDisposable
...
+
     public Vertex[] Vertices { get; protected set; }
 
     public Mesh(Vertex[] vertices, Vector3 position)
     {
+        createdVertexArrayObject = false;
+        this.vertices = vertices;
         Vertices = vertices;
-        CreateVertexArrayObject();
+        //CreateVertexArrayObject();
```
Comment (KarolDenst) [inline Hyper/Meshes/Mesh.cs:35]: uncomment or remove

File: Hyper/Scene.cs [hunk 11]
```diff
@@ -12,13 +13,16 @@
...
 using OpenTK.Windowing.GraphicsLibraryFramework;
+using System.Collections.Concurrent;
 
 
 namespace Hyper;
 
 internal class Scene : IInputSubscriber
 {
-    private readonly List<Chunk> _chunks;
+    public readonly ConcurrentDictionary<Guid, Chunk> _existingChunks;
```
Comment (KarolDenst) [inline Hyper/Scene.cs:23]: ExistingChunks

Comment (KarolDenst) [reply Hyper/Scene.cs:23]: Also mayber remove _existingChunks and add a method to get them from chunkWorker

File: Hyper/Scene.cs [hunk 12]
```diff
@@ -258,24 +239,29 @@ public void RegisterCallbacks()
 
         context.RegisterMouseButtonHeldCallback(MouseButton.Left, (e) =>
         {
-            foreach (var chunk in _chunks)
+            var target = Conversions.ToOpenTKVector(_player.GetCharacterRay(Camera.Front, 1));
+            
+            var closeChunks = _existingChunks.Values.Where(x => Chunk.MaxStraightDistance(x.Center, target) < Chunk.Size);
```
Comment (KarolDenst) [inline Hyper/Scene.cs:244]: add a method for this in chunkWorker

File: Hyper/MarchingCubes/ChunkWorker.cs [hunk 13]
```diff
@@ -0,0 +1,114 @@
+using System;
```
Comment (KarolDenst) [inline Hyper/MarchingCubes/ChunkWorker.cs:1]: there is a bunch of unused usings. Removed them from all classes

File: Hyper/MarchingCubes/ChunkWorker.cs (hunk 3 above)
Comment (mickamcia) [reply Hyper/MarchingCubes/ChunkWorker.cs:17]: I can easily change this to concurrent set. i was thinking about a situation where we maybe had a conflict between freshly generated chunk and a chunk stored in a database

File: Hyper/MarchingCubes/ChunkWorker.cs [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+            _neededChunkPositions.TryAdd(new Vector3i(0, 0, 0), Guid.NewGuid()); // initial chunk...
+            _existingChunkPositions = new ConcurrentDictionary<Vector3i, Guid>();
+            _existingChunks = existingChunks;
+            _scalarFieldGenerator = scalarFieldGenerator;
+            _chunkFactory = new ChunkFactory(_scalarFieldGenerator);
+            _simulationManager = simulationManager;
+        }
+        public void StartProcessing()
+        {
+            var processingThread = new Thread(ProcessItems);
```
Comment (KarolDenst) [inline Hyper/MarchingCubes/ChunkWorker.cs:38]: You can do this:
var processingThread = new Thread(ProcessItems)
            {
                IsBackground = true
            };

File: Hyper/MarchingCubes/ChunkWorker.cs (hunk 3 above)
Comment (mickamcia) [reply Hyper/MarchingCubes/ChunkWorker.cs:17]: image
The only problem is that there is no concurrent set to my knowledge.
https://github.com/dotnet/runtime/issues/16443
I only took a glance.

File: Hyper/MarchingCubes/ChunkWorker.cs [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+using System.Threading;
+using Hyper.Collisions;
+using Hyper.MarchingCubes;
+using Hyper.Meshes;
+using OpenTK.Mathematics;
+using Hyper.Collisions.Bepu;
+
+namespace Hyper
+{
+    internal class ChunkWorker
```
Comment (AleksyBalazinski) [inline Hyper/MarchingCubes/ChunkWorker.cs:14]: Have you considered using the [producer consumer pattern](https://jenkov.com/tutorials/java-concurrency/producer-consumer.html) here? It seems like a good place for it.
The producer thread would determine chunk visibility based on distance and send things like (remove, (0,1,0)), (add, (2, 0, 0)) to the message queue. Then the consumer would perform chunk addition/removal. I found implementation ideas here https://stackoverflow.com/questions/1656404/c-sharp-producer-consumer but it seems that there's something like that already provided by C# : https://learn.microsoft.com/en-us/dotnet/standard/collections/thread-safe/blockingcollection-overview?redirectedfrom=MSDN.

File: Hyper/MarchingCubes/ChunkWorker.cs (hunk 3 above)
Comment (AleksyBalazinski) [reply Hyper/MarchingCubes/ChunkWorker.cs:17]: If we don't use values at all, then could we just have `ConcurrentDictionary<Vector3i, int>` and aways add `(vec, 0)` for readability?

Comment (AleksyBalazinski) [reply Hyper/MarchingCubes/ChunkWorker.cs:17]: Just as a side note, I looked up if there's some sort of "throw away" type in C# that could be used here., like in C++ there's `nullptr_t`.  And it turns out that in C# the `null` literal has its own type! The only thing is "There is no way to explicitly write the null type" (https://stackoverflow.com/a/8204622/15767108)

File: Hyper/MarchingCubes/MeshGenerator.cs [hunk 16]
```diff
@@ -18,11 +18,11 @@ public MeshGenerator(Voxel[,,] scalarField, float isolevel = 0f)
     public Vertex[] GetMesh()
     {
         var vertices = new List<Vertex>();
-        for (int x = 0; x < _scalarField.GetLength(0) - 1; x++)
+        for (int x = 1; x < _scalarField.GetLength(0) - 2; x++)
```
Comment (AleksyBalazinski) [inline Hyper/MarchingCubes/MeshGenerator.cs:21]: why did we change these?

File: Hyper/Meshes/Chunk.cs [hunk 17]
```diff
@@ -83,12 +95,11 @@ public bool Build(Vector3 location, float deltaTime, float brushWeight, int radi
         var x = (int)location.X - Position.X;
         var y = (int)location.Y - Position.Y;
         var z = (int)location.Z - Position.Z;
-
-        if (x < 0 || y < 0 || z < 0
-            || x > Size - 1 || y > Size - 1 || z > Size - 1
-            || _voxels[x, y, z].Value >= 1f)
+        
+        if (x < -radius || y < -radius || z < -radius
```
Comment (AleksyBalazinski) [inline Hyper/Meshes/Chunk.cs:99]: why was that changed?

File: Hyper/Meshes/Chunk.cs [hunk 18]
```diff
@@ -83,12 +95,11 @@ public bool Build(Vector3 location, float deltaTime, float brushWeight, int radi
...
+            || x > Size - 1 + radius || y > Size - 1 + radius || z > Size - 1 + radius)
             return false;
-
+        bool flag = false;
```
Comment (AleksyBalazinski) [inline Hyper/Meshes/Chunk.cs:102]: change to something more descriptive

File: Hyper/Meshes/Chunk.cs [hunk 19]
```diff
@@ -98,36 +109,49 @@ public bool Build(Vector3 location, float deltaTime, float brushWeight, int radi
...
+        }
         simulation.Shapes.RemoveAndDispose(_shape, bufferPool);
         var mesh = MeshHelper.CreateMeshFromChunk(this, bufferPool);
         _shape = simulation.Shapes.Add(mesh);
         simulation.Statics[_handle].SetShape(_shape);
     }
+    public void DisposeCollisionSurface(Simulation simulation, BufferPool bufferPool)
+    {        
+        simulation.Statics.Remove(_handle);
+        simulation.Shapes.RemoveAndDispose(_shape, bufferPool); //datarace if in other thread
```
Comment (AleksyBalazinski) [inline Hyper/Meshes/Chunk.cs:141]: do these data races occur? If so, we need to fix that

Comment (AleksyBalazinski) [reply Hyper/Meshes/Chunk.cs:141]: I don't know if that's related or not, but when mining/building I'm getting `InvalidOperation` messages in the console. This needs to be investigated

Comment (AleksyBalazinski) [reply Hyper/Meshes/Chunk.cs:141]: When building we can fall through the map. I think this should be solved in this PR

File: Hyper/Meshes/Mesh.cs [hunk 20]
```diff
@@ -20,18 +23,27 @@ internal class Mesh : IDisposable
 
     protected int NumberOfVertices;
 
+    private bool createdVertexArrayObject;
+
     public Vertex[] Vertices { get; protected set; }
 
     public Mesh(Vertex[] vertices, Vector3 position)
     {
+        createdVertexArrayObject = false;
```
Comment (AleksyBalazinski) [inline Hyper/Meshes/Mesh.cs:32]: once this variable is set to true, it can never get back to false again - do we need it in this case?

File: Hyper/Meshes/Chunk.cs [hunk 21]
```diff
@@ -25,8 +25,14 @@ internal class Chunk : Mesh
 
     private TypedIndex _shape;
 
+    private BepuPhysics.Collidables.Mesh? collisionSurface;
```
Comment (AleksyBalazinski) [inline Hyper/Meshes/Chunk.cs:28]: is there a way to get rid of nullable values? They generally make the code less robust

File: Hyper/Meshes/Chunk.cs (hunk 7 above)
Comment (mickamcia) [reply Hyper/Meshes/Chunk.cs:138]: usages are commented out because of an existing issue with the physics engine

File: Hyper/Meshes/Mesh.cs [hunk 22]
```diff
@@ -10,6 +10,9 @@ internal class Mesh : IDisposable
 {
     public Vector3 Position { get; set; }
 
+    public Vertex[] vertices;
+    //Will also have to add rotation and scale
```
Comment (AleksyBalazinski) [inline Hyper/Meshes/Mesh.cs:14]: is this still relevant? We have `Scaling` and `RigidPose` which has rotation

File: Hyper/Meshes/Mesh.cs [hunk 23]
```diff
@@ -52,7 +69,7 @@ public virtual void RenderFullDescription(Shader shader, float scale, Vector3 ca
         GL.DrawArrays(PrimitiveType.Triangles, 0, NumberOfVertices);
     }
 
-    private void CreateVertexArrayObject()
+    public void CreateVertexArrayObject()
```
Comment (AleksyBalazinski) [inline Hyper/Meshes/Mesh.cs:72]: can it stay private? It's only referenced from its own class

**Review by AleksyBalazinski — CHANGES_REQUESTED** [review]
I left some general thoughts in the comments, but aside from those, maybe we could also look into some performance problems here? Specifically, on my machine (in debug mode) building/mining causes the fps to drop 3x

File: Hyper/Meshes/Mesh.cs (hunk 8 above)
Comment (mickamcia) [reply Hyper/Meshes/Mesh.cs:13]: the whole field was unnecessary

File: Hyper/Scene.cs (hunk 11 above)
Comment (mickamcia) [reply Hyper/Scene.cs:23]: _existingChunks is useful there is useful when Mine and Build methods are called

File: Hyper/MarchingCubes/ChunkWorker.cs (hunk 1 above)
Comment (mickamcia) [reply Hyper/MarchingCubes/ChunkWorker.cs:78]: fixed with a semaphore

File: Hyper/Meshes/Chunk.cs (hunk 17 above)
Comment (mickamcia) [reply Hyper/Meshes/Chunk.cs:99]: this is because we want to take into consideration a position that we want to mine that is outside of the chunk. this fixes the problem of mining/building near the edges, but is also a bit more computationally expensive

File: Hyper/Meshes/Chunk.cs (hunk 19 above)
Comment (mickamcia) [reply Hyper/Meshes/Chunk.cs:141]: Basically i don't know how to remove a CollisionSurface... if i do it in a way provided above, the simulation often tries to use it resulting in a null exception. Right now these surfaces stay indefinitely, which is obviously an issue.

Comment (mickamcia) [reply Hyper/Meshes/Chunk.cs:141]: I do not think falling through the map is an issue on my part, or issue at all. I dramatically increased the volume and strength of building/mining. Because of that, when we mine/build, a large surface of terrain shifts its position in a single tick and sometimes the surface "teleports" on the other side of the character. Then we fall...

File: Hyper/Meshes/Mesh.cs (hunk 22 above)
Comment (mickamcia) [reply Hyper/Meshes/Mesh.cs:14]: i do not think the comment is mine, vertices field is in fact completely unnecessary

File: Hyper/Meshes/Mesh.cs (hunk 9 above)
Comment (mickamcia) [reply Hyper/Meshes/Mesh.cs]: opengl does not like nullables

**Comment by KarolDenst** [comment]
If we uncomment the line with DisposeCollisionSurface sometimes DisposeCollisionSurface is called before CreateCollisionSurface which causes the app to crush.

File: Hyper/MarchingCubes/ChunkWorker.cs (hunk 1 above)
Comment (KarolDenst) [reply Hyper/MarchingCubes/ChunkWorker.cs]: the issue still exists but only on debug mode

File: Hyper/MarchingCubes/ChunkWorker.cs (hunk 15 above)
Comment (mickamcia) [reply Hyper/MarchingCubes/ChunkWorker.cs:14]: you are totally right

File: Hyper/Meshes/Chunk.cs (hunk 21 above)
Comment (mickamcia) [reply Hyper/Meshes/Chunk.cs]: i mean we can, i could use something similar to createdVertexArrayObject flag, which isn't great-looking. give me a hint if you know a professional way of dealing with such problems
